File: ankiconnect-server/src/services/cardExtraction.ts
import * as ankiService from './anki.js';
import { getSettings } from './settings.js';
import type { AnkiNote, CardContent, CardPreview } from 'shared';

export interface CardResponse {
  word: string;
  definition: string;
  exampleSentence: string;
  sentenceTranslation: string;
  rootWord?: string;
  spellingCorrection?: string;
}

/**
 * Search Anki for a word, storing the full note (or null).
 * Silently stores null on errors (Anki may be offline).
 */
async function checkAnki(
  word: string,
  deck: string,
  results: Map<string, AnkiNote | null>
): Promise<void> {
  if (results.has(word)) return;
  try {
    const note = await ankiService.searchWordCached(word, deck);
    results.set(word, note);
  } catch (error) {
    console.warn('[CardExtraction] Anki search failed for "%s":', word, error);
    results.set(word, null);
  }
}

/**
 * Reverse-map an AnkiNote's fields to CardContent using the field mapping from settings.
 */
function noteToCardContent(note: AnkiNote, fieldMapping: Record<string, string>): CardContent {
  // Build reverse mapping: model field name → standard field name
  // e.g. { Bangla: "Word", Eng_trans: "Definition", ... }
  const reverseMap = new Map<string, string>();
  for (const [standard, modelField] of Object.entries(fieldMapping)) {
    reverseMap.set(modelField, standard);
  }

  const getField = (standardName: string): string => {
    // Try mapped field name first, then standard name as fallback
    const mappedName = fieldMapping[standardName] || standardName;
    return note.fields[mappedName]?.value || '';
  };

  return {
    word: getField('Word'),
    definition: getField('Definition'),
    exampleSentence: getField('Example'),
    sentenceTranslation: getField('Translation'),
  };
}

/**
 * Build CardPreview[] from parsed AI card responses + Anki duplicate check results.
 */
export async function buildCardPreviews(
  cards: CardResponse[],
  targetDeck: string,
  ankiResults: Map<string, AnkiNote | null>
): Promise<CardPreview[]> {
  // Check Anki for any words we haven't checked yet
  await Promise.all(
    cards
      .filter((card) => !ankiResults.has(card.word))
      .map((card) => checkAnki(card.word, targetDeck, ankiResults))
  );

  const settings = await getSettings();
  const fieldMapping = settings.fieldMapping || {};

  return cards.map((card) => {
    const existingNote = ankiResults.get(card.word);
    const existingCard =
      existingNote && existingNote.noteId !== 0
        ? noteToCardContent(existingNote, fieldMapping)
        : undefined;

    return {
      word: card.word,
      definition: card.definition,
      exampleSentence: card.exampleSentence,
      sentenceTranslation: card.sentenceTranslation,
      rootWord: card.rootWord,
      spellingCorrection: card.spellingCorrection,
      alreadyExists: !!existingNote,
      existingCard,
    };
  });
}

File: python-server/anki_defs/services/card_extraction.py
"""Card extraction from AI responses — validation, Anki checking, preview building.

Ports Express cardExtraction.ts + addon card_extraction.py into shared service.
"""

from __future__ import annotations

import re
from typing import Any


def validate_card_responses(parsed: Any) -> list[dict[str, Any]]:
    """Validate and coerce a parsed AI response into card dicts. Raises if unrecoverable."""
    items = parsed if isinstance(parsed, list) else [parsed]
    cards = []
    for i, item in enumerate(items):
        if not isinstance(item, dict):
            raise ValueError(f"Card {i}: expected object, got {type(item).__name__}")
        word = str(item.get("word", "")).strip()
        if not word:
            raise ValueError(f'Card {i}: missing "word" field')
        card: dict[str, Any] = {
            "word": word,
            "definition": str(item.get("definition", "")),
            "nativeDefinition": str(item.get("nativeDefinition", "")),
            "exampleSentence": str(item.get("exampleSentence", "")),
            "sentenceTranslation": str(item.get("sentenceTranslation", "")),
        }
        spelling = item.get("spellingCorrection")
        if spelling:
            card["spellingCorrection"] = str(spelling)
        cards.append(card)
    return cards


def apply_spelling_correction(sentence: str, correction: str) -> str:
    """Parse 'input → standard' and apply to sentence, preserving **bold** markers."""
    match = re.match(r"^(.+?)\s*→\s*(.+)$", correction)
    if not match:
        return sentence
    wrong = match.group(1)
    right = match.group(2)
    result = sentence.replace(wrong, right).replace(f"**{wrong}**", f"**{right}**")
    return result


def _note_to_card_content(
    note: dict[str, Any], field_mapping: dict[str, str]
) -> dict[str, str]:
    """Reverse-map an Anki note's fields to card content using the field mapping."""

    def get_field(standard_name: str) -> str:
        mapped_name = field_mapping.get(standard_name, standard_name)
        field_data = note.get("fields", {}).get(mapped_name)
        if field_data:
            return field_data.get("value", "")
        return ""

    return {
        "word": get_field("Word"),
        "definition": get_field("Definition"),
        "nativeDefinition": get_field("NativeDefinition"),
        "exampleSentence": get_field("Example"),
        "sentenceTranslation": get_field("Translation"),
    }


def build_card_previews(
    cards: list[dict[str, Any]],
    target_deck: str,
    anki_results: dict[str, Any | None],
    field_mapping: dict[str, str],
) -> list[dict[str, Any]]:
    """Build card preview dicts from parsed AI JSON cards + Anki duplicate checks."""
    previews = []
    for card in cards:
        word = card.get("word", "")
        existing_note = anki_results.get(word)

        existing_card = None
        if existing_note and existing_note.get("noteId", 0) != 0:
            existing_card = _note_to_card_content(existing_note, field_mapping)

        example_sentence = card.get("exampleSentence", "")
        spelling_correction = card.get("spellingCorrection")
        if spelling_correction and example_sentence:
            example_sentence = apply_spelling_correction(example_sentence, spelling_correction)

        preview: dict[str, Any] = {
            "word": word,
            "definition": card.get("definition", ""),
            "nativeDefinition": card.get("nativeDefinition", ""),
            "exampleSentence": example_sentence,
            "sentenceTranslation": card.get("sentenceTranslation", ""),
            "alreadyExists": existing_note is not None,
        }
        if spelling_correction:
            preview["spellingCorrection"] = spelling_correction
        if existing_card:
            preview["existingCard"] = existing_card

        previews.append(preview)

    return previews
